src/docfx/docset/Context.cs

```diff
@@ -1,23 +1,31 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Docs.Build
 {
     internal class Context
     {
         private readonly string _outputPath;
         private readonly Report _report;
+        private readonly ConcurrentDictionary<string, Lazy<(List<Error>, JToken)>> _cache;
 
         public Context(Report report, string outputPath)
         {
             _report = report;
             _outputPath = Path.GetFullPath(outputPath);
+            _cache = new ConcurrentDictionary<string, Lazy<(List<Error>, JToken)>>();
         }
 
+        public (List<Error> errors, JToken token) Load(Document file, Func<Document, (List<Error>, JToken)> action)
+            => _cache.GetOrAdd(file.FilePath + file.LastWriteTime, new Lazy<(List<Error>, JToken)>(() => action(file))).Value;
```

Comment:
this is a two-way cache, we can make the cache 1 way in and 1 way out? like xxxbuilder

Reply:
Done :)

---
In reply to: [221175823](https://github.com/dotnet/docfx/pull/3458#discussion_r221175823) [](ancestors = 221175823)